'''
exceptions.py: exceptions defined by Zuppa

Authors
-------

Michael Hucka <mhucka@caltech.edu> -- Caltech Library

Copyright
---------

Copyright (c) 2020 by Michael Hucka and the California Institute of Technology.
This code is open-source software released under a 3-clause BSD license.
Please see the file "LICENSE" for more information.
'''


# Base class.
# .............................................................................
# The base class makes it possible to use a single test to distinguish between
# exceptions generated by Zuppa and exceptions generated by something else.

class ZuppaException(Exception):
    '''Base class for Zuppa exceptions.'''
    pass


# Exception classes.
# .............................................................................

class CannotProceed(ZuppaException):
    '''A recognizable condition caused an early exit from the program.'''
    pass

class UserCancelled(ZuppaException):
    '''The user elected to cancel/quit the program.'''
    pass

class NetworkFailure(ZuppaException):
    '''Unrecoverable problem involving network operations.'''
    pass

class ServiceFailure(ZuppaException):
    '''Unrecoverable problem involving a remote service.'''
    pass

class AuthenticationFailure(ZuppaException):
    '''Problem obtaining or using authentication credentials.'''
    pass

class InternalError(ZuppaException):
    '''Unrecoverable problem involving eprints2bags itself.'''
    pass

class NotFound(ZuppaException):
    '''Item could not be found.'''
    pass
